File: Java Advanced/3. Sets and Maps/Sets and Maps/src/PopulationCounter.java
import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;

public class PopulationCounter {
    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        Map<String, Map<String, Long>> map = new LinkedHashMap<>();
        String line;

        while (!"report".equals(line = s.nextLine())) {

            String[] tokens = line.split("\\|");
            String city = tokens[0];
            String country = tokens[1];
            long population = Long.parseLong(tokens[2]);

            Map<String, Long> helper = new LinkedHashMap<>();
            helper.put(city, population);
            if (!map.containsKey(country)) {
                map.put(country, helper);
            } else {
                map.get(country).put(city, population);
            }
        }

        Map<String, Map<String, Long>> sortedMap = map.entrySet().stream()
                .sorted((a, b) -> {
                    long populatrion1 = a.getValue()
                            .values()
                            .stream()
                            .mapToLong(Long::valueOf)
                            .sum();
                    long populatrion2 = b.getValue()
                            .values()
                            .stream()
                            .mapToLong(Long::valueOf)
                            .sum();
                    return Long.compare(populatrion2, populatrion1);
                }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (a, b) -> a, LinkedHashMap::new));

        for (Map.Entry<String, Map<String, Long>> entry : sortedMap.entrySet()) {
            long totalPopulation = 0;
            Map<String, Long> cities = entry.getValue().entrySet()
                    .stream()
                    .sorted((a, b) -> Long.compare(b.getValue(), a.getValue()))
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (a, b) -> a, LinkedHashMap::new));
            StringBuilder temp = new StringBuilder();
            for (Map.Entry<String, Long> city : cities.entrySet()) {
                totalPopulation += city.getValue();
                temp.append("=>").append(city.getKey()).append(": ").append(city.getValue()).append("\n");

            }
            System.out.printf("%s (total population: %d)%n", entry.getKey(), totalPopulation);
            System.out.print(temp);
        }

        System.out.println();
    }
}

File: soft uni prgramming fundamentals/7. Dictionaries, Lambda Expressions and LINQ/Dictionaries/PopulationCounter/PopulationCounter.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace PopulationCounter
{
    class PopulationCounter
    {
        static void Main(string[] args)
        {
            string[] intput = Console.ReadLine().Split('|').ToArray();
            Dictionary<string, Dictionary<string, long>> report = new Dictionary<string, Dictionary<string, long>>();
            while (intput[0] != "report")
            {
                string city=intput[0];
                string country=intput[1];
                int popolation= int.Parse(intput[2]);

                if (report.ContainsKey(country) == false)
                {
                    Dictionary<string, long> cities = new Dictionary<string, long>();
                    if (cities.ContainsKey(city) == false)
                    {
                        cities.Add(city, popolation);
                    }
                    else
                    {
                        cities[city] = popolation;
                    }
                    report.Add(country, cities);
                }
                else
                {
                    if(report[country].ContainsKey(city)==false)
                    {
                        report[country].Add(city, popolation);
                    }
                   
                    
                }
                intput = Console.ReadLine().Split('|').ToArray();

            }
            foreach (var count in report.OrderByDescending(x=>x.Value.Values.Sum()) )
            {
                Console.WriteLine(value: $"{count.Key} (total population: {count.Value.Values.Sum()})");
                foreach (var item in count.Value.OrderByDescending(x=>x.Value))
                {
                    Console.WriteLine($"=>{item.Key}: {item.Value}");
                }
            }
        }
    }
}
